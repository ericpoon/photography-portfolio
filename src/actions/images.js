import firebase from '../firebase/firebase';

const database = firebase.database();

export const setImages = images => ({
  type: 'SET_IMAGES',
  images,
});

export const startDownloadImages = () => {
  return (dispatch) => {
    return database.ref('/images').once('value')
      .then((snapshot) => {
        const images = [];
        snapshot.forEach((childSnapshot) => {
          images.push({
            id: childSnapshot.key,
            ...childSnapshot.val(),
          });
        });
        dispatch(setImages(images));
      });
  };
};

export const addImage = image => ({
  type: 'ADD_IMAGE',
  image,
});

export const startAddImage = (image) => {
  return (dispatch) => {
    return database.ref('/images').push(image)
      .then((ref) => {
        dispatch(addImage({
          id: ref.key,
          ...image,
        }));
      });
  };
};

export const editImage = (id, updates) => ({
  type: 'EDIT_IMAGE',
  id,
  updates,
});

export const startEditImage = (id, updates) => {
  return (dispatch) => {
    return database.ref(`/images/${id}`).update(updates)
      .then(() => {
        dispatch(editImage(id, updates));
      });
  };
};

export const deleteImage = id => ({
  type: 'DELETE_IMAGE',
  id,
});

export const startDeleteImage = (id) => {
  return (dispatch) => {
    return database.ref(`images/${id}`).remove()
      .then(() => {
        dispatch(deleteImage(id));
      });
  };
};
