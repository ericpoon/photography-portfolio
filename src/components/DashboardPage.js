import React from 'react';

export default () => (
  <div>
    <p>Dashboard page goes here</p>
  </div>
);
