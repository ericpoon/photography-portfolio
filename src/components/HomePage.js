import React from 'react';

const HomePage = (props) => {
  const { history } = props;
  const onEnterClick = () => history.push('/portfolio');
  const onLoginClick = () => history.push('/login');

  return (
    <div className={'banner-layout'}>
      <div className={'banner-layout__banner'}>
        <h1 className={'banner-layout__subtitle'}>Firstname Lastname</h1>
        <p className={'banner-layout__title'}>Hong Kong</p>
      </div>
      <div>
        <button onClick={onEnterClick} className={'button button--larger'}>Enter</button>
      </div>
      <button onClick={onLoginClick} className={'button button--bottom-right'}>Login</button>
    </div>
  );
};

export default HomePage;
