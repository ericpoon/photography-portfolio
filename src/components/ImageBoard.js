import React from 'react';
import { DragDropContext } from 'react-dnd';
import HTML5Backend from 'react-dnd-html5-backend';
import SortableContainer from './SortableContainer';

const ROW_NUM = 3;
const COL_NUM = 3;

// todo: Item states are not persisted,
// use redux store to keep track of those images being edited (not all)

class ImageBoard extends React.Component {
  constructor(props) {
    super();
    this.state = { images: props.children };
  }

  componentWillReceiveProps(nextProps, nextContext) {
    this.setState({ images: nextProps.children });
  }

  moveCard = (dragIndex, hoverIndex) => {
    const { images } = this.state;
    const draggedCard = images[dragIndex];
    images.splice(dragIndex, 1, images[hoverIndex]);
    images.splice(hoverIndex, 1, draggedCard);
    this.setState({ images });
  };

  render() {
    const { images } = this.state;
    const { sortable } = this.props;
    const grid = [];

    images.forEach((current, index) => {
      const row = Math.floor(index / ROW_NUM);
      const col = index % COL_NUM;
      if (!grid[row]) grid[row] = [];
      grid[row][col] = sortable ? (
        <SortableContainer
          moveCard={this.moveCard}
          index={index}
          key={current.key}
        >
          {current}
        </SortableContainer>
      ) : current;
    });

    return (
      <div className={'image-card-collection'}>
        {grid.map(row => (
          <div className={'image-card-collection__row'} key={row.map(i => i.key).toString()}>
            {row.map(individual => individual)}
          </div>
        ))}
      </div>
    );
  }
}

export default DragDropContext(HTML5Backend)(ImageBoard);
