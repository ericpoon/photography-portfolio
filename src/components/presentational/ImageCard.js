import React from 'react';

const ImageCard = (props) => {
  const {
    title, subtitle, description, fileUrl, onImageClick,
  } = props;

  return (
    <div className={'image-card'}>
      <div
        className={'image-card__image'}
        style={{ backgroundImage: `url(${fileUrl})` }}
        onClick={onImageClick}
      />
      <div>
        <h3 className={'image-card__title'}>{title}</h3>
        <h4 className={'image-card__subtitle'}>{subtitle}</h4>
        <p className={'image-card__description'}>{description}</p>
      </div>
    </div>
  );
};

export default ImageCard;
