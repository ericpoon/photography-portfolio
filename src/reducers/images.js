const initialState = [];

export default (state = initialState, action) => {
  switch (action.type) {
    case 'SET_IMAGES':
      return action.images;
    case 'ADD_IMAGE':
      return [...state, action.image];
    case 'EDIT_IMAGE':
      return state.map((i) => {
        return i.id === action.id ? { ...i, ...action.updates } : i;
      });
    case 'DELETE_IMAGE':
      return state.filter(i => i.id !== action.id);
    default:
      return state;
  }
};
