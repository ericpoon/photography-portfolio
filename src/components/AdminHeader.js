import React from 'react';
import { connect } from 'react-redux';
import Header from './presentational/Header';
import { startLogout } from '../actions/auth';

export const AdminHeader = (props) => {
  return (
    <Header title={'Portfolio'}>
      <button
        onClick={() => null}
        className={'button button--no-background'}
      >
        {props.userName}
      </button>

      <button
        onClick={props.startLogout}
        className={'button button--no-background'}
      >
        Logout
      </button>
    </Header>
  );
};

const mapStateToProps = state => ({
  userName: state.auth.name,
});

export default connect(mapStateToProps, { startLogout })(AdminHeader);
