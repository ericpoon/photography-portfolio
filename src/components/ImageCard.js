import React from 'react';
import { connect } from 'react-redux';
import { addImage, editImage, deleteImage } from '../actions/images';

export class ImageCard extends React.Component {
  constructor(props) {
    super();
    const { title, subtitle, description } = props;
    this.state = { title, subtitle, description };
  }

  onTitleChange = (e) => {
    this.setState({ title: e.target.value.trim() });
  };
  onSubtitleChange = (e) => {
    this.setState({ subtitle: e.target.value.trim() });
  };
  onDescriptionChange = (e) => {
    this.setState({ description: e.target.value.trim() });
  };
  onUpdateClick = () => {
    const { title, subtitle, description } = this.state;
    const updates = { title, subtitle, description };
    this.props.editImage(this.props.id, updates);
  };
  onDiscardClick = () => {
    const { title, subtitle, description } = this.props;
    this.setState({ title, subtitle, description });
  };
  onRemoveClick = () => {
    this.props.deleteImage(this.props.id);
  };

  render() {
    const { onImageClick, editable } = this.props;
    const { title, subtitle, description } = this.state;
    if (editable) {
      return (
        <div className={'image-card'}>
          <div className={'image-card__image'} onClick={onImageClick}>
            <button
              className={'button button--floating-top-left button--danger'}
              onClick={this.onRemoveClick}
            />
          </div>
          <div>
            <input
              className={
                `image-card__title image-card__title--input${
                  this.state.title !== this.props.title ? ' is-dirty' : ''}`
              }
              value={title}
              onChange={this.onTitleChange}
            />
            <input
              className={
                `image-card__subtitle image-card__subtitle--input${
                  this.state.subtitle !== this.props.subtitle ? ' is-dirty' : ''}`
              }
              value={subtitle}
              onChange={this.onSubtitleChange}
            />
            <textarea
              className={
                `image-card__description image-card__description--textarea${
                  this.state.description !== this.props.description ? ' is-dirty' : ''}`
              }
              value={description}
              onChange={this.onDescriptionChange}
            />
            <div className={'image-card__actions'}>
              <button className={'button'} onClick={this.onUpdateClick}>Update</button>
              <button className={'button'} onClick={this.onDiscardClick}>Discard</button>
            </div>
          </div>
        </div>
      );
    }

    return (
      <div className={'image-card'}>
        <div className={'image-card__image'} onClick={onImageClick} />
        <div>
          <h3 className={'image-card__title'}>{title}</h3>
          <h4 className={'image-card__subtitle'}>{subtitle}</h4>
          <p className={'image-card__description'}>{description}</p>
        </div>
      </div>
    );
  }
}

export default connect(undefined, { addImage, editImage, deleteImage })(ImageCard);
