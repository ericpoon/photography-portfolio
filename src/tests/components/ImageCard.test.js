import React from 'react';
import { shallow } from 'enzyme';
import { ImageCard } from '../../components/ImageCard';
import images from '../fixtures/images';

let wrapper,
  title,
  subtitle,
  description,
  onImageClick,
  addImage,
  editImage,
  deleteImage;

beforeEach(() => {
  const { title: t, subtitle: s, description: d } = images[0];
  title = t;
  subtitle = s;
  description = d;
  onImageClick = jest.fn();
  addImage = jest.fn();
  editImage = jest.fn();
  deleteImage = jest.fn();
  wrapper = shallow(<ImageCard
    editable={false}
    title={title}
    subtitle={subtitle}
    description={description}
    onImageClick={onImageClick}
    addImage={addImage}
    editImage={editImage}
    deleteImage={deleteImage}
  />);
});

describe('non-editable/view-only ImageCard', () => {
  it('should render non-editable ImageCard correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should handle onImageClick', () => {
    wrapper.find('div>div').at(0).prop('onClick')();
    expect(onImageClick).toHaveBeenCalled();
  });
});

describe('editable ImageCard', () => {
  beforeEach(() => {
    wrapper.setProps({ editable: true });
  });

  it('should render editable ImageCard correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should handle onImageClick', () => {
    wrapper.find('div>div').at(0).prop('onClick')();
    expect(onImageClick).toHaveBeenCalled();
  });

  it('should handle delete image', () => {
    wrapper.find('button').at(0).simulate('click');
    expect(deleteImage).toHaveBeenCalled();
  });

  it('should handle edit image', () => {
    wrapper.find('button').at(1).simulate('click');
    expect(editImage).toHaveBeenCalled();
  });

  it('should handle discard changes', () => {
    wrapper.setState({ title: 'edited title', subtitle: 'edited subtitle', description: 'edited description' });
    wrapper.find('button').at(2).simulate('click');
    expect(wrapper.state()).toMatchObject({ title, subtitle, description });
  });
});
