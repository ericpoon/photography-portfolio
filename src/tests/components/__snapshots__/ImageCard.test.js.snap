// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`editable ImageCard should render editable ImageCard correctly 1`] = `
<div
  className="image-card"
>
  <div
    className="image-card__image"
    onClick={[MockFunction]}
  >
    <button
      className="button button--floating-top-left button--danger"
      onClick={[Function]}
    />
  </div>
  <div>
    <input
      className="image-card__title image-card__title--input"
      onChange={[Function]}
      value="Kowloon"
    />
    <input
      className="image-card__subtitle image-card__subtitle--input"
      onChange={[Function]}
      value="Dec 31, 2017"
    />
    <textarea
      className="image-card__description image-card__description--textarea"
      onChange={[Function]}
      value="Putarem quodque tamquam ii ob an deumque. Me assequi sapores ab verarum. Mei qui religionis nam fas devenietur ponderibus. Junctae invenit plausum differt de ipsamet externo is. Mutationum sui negationem exponuntur"
    />
    <div
      className="image-card__actions"
    >
      <button
        className="button"
        onClick={[Function]}
      >
        Update
      </button>
      <button
        className="button"
        onClick={[Function]}
      >
        Discard
      </button>
    </div>
  </div>
</div>
`;

exports[`non-editable/view-only ImageCard should render non-editable ImageCard correctly 1`] = `
<div
  className="image-card"
>
  <div
    className="image-card__image"
    onClick={[MockFunction]}
  />
  <div>
    <h3
      className="image-card__title"
    >
      Kowloon
    </h3>
    <h4
      className="image-card__subtitle"
    >
      Dec 31, 2017
    </h4>
    <p
      className="image-card__description"
    >
      Putarem quodque tamquam ii ob an deumque. Me assequi sapores ab verarum. Mei qui religionis nam fas devenietur ponderibus. Junctae invenit plausum differt de ipsamet externo is. Mutationum sui negationem exponuntur
    </p>
  </div>
</div>
`;
